jre_emul/Classes/com/google/j2objc/security/cert/IosX509Certificate.java

```diff
@@ -142,107 +184,139 @@ public void verify(PublicKey key, String sigProvider)
     throw new AssertionError("not implemented");
   }
 
-  // The X509 certificate properties are not available from the iOS Security
-  // Framework API. To get this properties would require an ASN.1 decoder, so
-  // only implement these when they are required.
+  // The X509 certificate properties are indirectly available from the iOS Security
+  // Framework API. To get this properties would require an ASN.1 decoder
+  // We have these things since commit fa9c48a1 and 592382e0
+  // by using harmony's TBSCertificate, Certificate, etc.
 
+  // #getPublicKey#getEncoded can be used for public key pinning
   @Override
   public PublicKey getPublicKey() {
-    return null;
+    lazyDecoding();
+    return tbsCert.getSubjectPublicKeyInfo().getPublicKey();
   }
 
   @Override
   public BigInteger getSerialNumber() {
-    return null;
+    lazyDecoding();
+    return tbsCert.getSerialNumber();
   }
 
   @Override
   public Set<String> getCriticalExtensionOIDs() {
-    return null;
+    lazyDecoding();
+    return extensions.getCriticalExtensions();
   }
 
   @Override
   public byte[] getExtensionValue(String oid) {
-    return null;
+    lazyDecoding();
+    Extension ext = extensions.getExtensionByOID(oid);
+    return (ext == null) ? null : ext.getRawExtnValue();
   }
 
   @Override
   public Set<String> getNonCriticalExtensionOIDs() {
-    return null;
+    lazyDecoding();
+    return extensions.getNonCriticalExtensions();
   }
 
   @Override
   public boolean hasUnsupportedCriticalExtension() {
-    return false;
+    lazyDecoding();
+    return extensions.hasUnsupportedCritical();
   }
 
   @Override
   public int getVersion() {
-    return 0;
+    lazyDecoding();
+    return tbsCert.getVersion() + 1; // TODO determine why +1
```

Comment:
Remove the +1, as version numbers are zero-based, as described in section 4.1 of http://www.ietf.org/rfc/rfc3280.txt.